manpages/frommrc.man
.na
.nh
.TH FROMMRC 1 2.00 BL3DFS
.SH NAME
frommrc \- Converts a mrc file to SGI rgb image files.
.SH SYNOPSIS
frommrc [mrc filename] [rgb filename]
.SH DESCRIPTION
A series of rgb files will be created with the prefix 
[rgb filename] and with the suffix nnn.rgb,
where nnn is the z number.
.PP
If z for the mrc file is equal to 1 no suffix is
added to the rgb filename.
.PP
The image is automatically scaled so that the minimum
and maximum are 0 and 255 respectivily if the image
is more than 8-bits deep.
.PP
Color MRC images are translated directly.  The color
format for MRC is not standard.  The mode number
is 16 and the data is stored as three bytes;
red, green and blue.


.SH AUTHOR
Jim Kremer 

.SH SEE ALSO
rgbv    View SGI rgb files.
.br
tomrc   Convert MRC image files to SGI rgb image files.

.SH BUGS
Email bug reports to mast@colorado.edu.


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10/8/03: Modified PIP to read newly invented "autodoc" files to determine
program options and generate manual pages sections.

10/6/03: IMOD_3-0-22
10/6/03: Fixed the default number of views value for transferfid; it was not
being initialized in etomo.

10/6/03: Reordered thread control to eliminate race condition that often 
prevented matchvol1 from running right after any preceeding com script.

10/2/03: IMOD_3-0-21
10/2/03: Modified Rotatevol, Matchvol, and Warpvol to work on volumes up to 
16K in any dimension; Rotatevol was failing for an output X/Y size of 4K by 4K.

10/1/03: Made imodinfo work properly for counting scattered points on subsets
of the volume.

10/1/03: Added message for opening bead fixer window.

10/1/03: Added ability to turn a plugin into an internal module that is added
to the Special menu, to provide Bead fixer on Windows.

9/29/03: Restructured some 3dmod include files to restore ability to build
plugins with minimal header files.

9/25/03: Made sure windows stayed on the screen when restoring Zap and Info
window positions.

9/25/03: Changed Midas to number sections from 1 instead of 0.

9/24/03: IMOD_3-0-20
9/24/03: Resolved problems with the Fortran graphics package so it works
on all systems, at least briefly.

9/24/03: Fixed some problems with event sending under Windows, added a
retry capability to make it work reliably on the Mac, and found that it
did not work with the multi-threaded Qt library on the SGI.

9/24/03: Used different window properties for saving and restoring window
positions, which fixed problems in Redhat 7.3 and Irix.  Made it not restore
Info window size since that is highly font-dependent.

9/23/03: Put Fortran graphics programs inside application directories on the
Mac, which made them work correctly, with the window raisable and resizable.

9/22/03: Removed a large unused common from smoothsurf, which was making it
fail under Cygwin.

9/19/03: Made sample.com files produce top.rec, etc, directly as the output
of Tilt to avoid replacing existing .rec files.

9/19/03: Made 3dmod require only a 15-bit RGB visual, not 16 bit.

9/18/03: IMOD_3-0-19
9/18/03: Added the ability to float (match contrast between sections) based
on a subarea, either the image displayed or the area inside of a rubberband
in the active Zap window.  Also added an autocontrast button to the Info
window with target contrasts settable from the options dialog.

9/17/03: Made the Zap window pan by one display pixel per pixel of mouse move
for zooms less than one.

9/16/03: Made 3dmod save and restore sizes and positions of the Zap and Info
windows, with different settings saved for different image sizes.  Zoom is
adjusted to fit the whole image to the window.  If there are no application
saved settings, the Zap window now opens up wit a size that fits a zoomed-down
image if appropriate.

9/15/03: Eliminated physical flipping of data in 3dmod to avoid very bad
performance when the image data gets into swap space.

9/15/03: Optimized image drawing for zooms equal to the inverse of an integer.

9/13/03: IMOD_3-0-18
9/12/03: Prevented buffer overruns in an array that holds the filename in
3dmod, fixed problems with model checksums that were leading to queries to
save an unchanged model, arranged the startup page in two columns, fixed a
problem with zooming up under Windows, and added a workaround to a display
problem with Qt 3.2.1 on the Mac.

9/11/03: IMOD_3-0-17  beta release with Etomo

9/9/03: IMOD_3-0-16
9/5/03: Added a program, Resamplemod, for rotating a microtubule model to be
perpendicular to the Z axis and creating points at each section or plane
along that axis.

9/4/03: Incorporated changes in Mtlengths to correct for oblique microtubules
in tilted sections and implemented PIP input.

8/29/03: Set up and documented installation procedures on the Mac.

8/29/03: Changed the Fortran graphics routines to run the graphics window in
a separate thread from the Fortran program.  This only works properly under
Linux, where the window refreshes itself when resized or exposed.  On the Mac,
the window can now be moved but not resized.

8/25/03: Added the -E option to Imodmesh to cap the ends of tubes.

8/13/03: IMOD_3-0-15
8/12/03: Made the Qt-based graphics routines used in Fortran programs work on
the Mac without hanging when the window opened.

8/12/03: Made it possible to start up Nda without a model file, just like Mtk.

8/9/03: IMOD_3-0-14
8/9/03: Changed Blendmont to take PIP input and modified command files and
Copytomocoms to use new form of input.

8/8/03: Etomo was incorporated into IMOD build structure.

8/8/03: Wrote man page for Mtk, changed program to make terminology
consistent, to fix problems with saving connector objects and moving items to
new objects, and to add ability to export a graph in a format easily readable
by other programs.  (Nda can also export graphs.)  Also provided some example
command files in the com directory.

8/8/03: Fixed problem of current point not being displayed when model is
turned off, caused by the ability to display the vector from bead fixer.

8/1/03: Made it possible to kill 3dmod during image flipping operation.

8/1/03: Standardized error exit in Xfproduct.

8/1/03: Implemented proper thickness padding in fast backprojection code of
Tilt, reduced Tilt's stack size so the increase is only 3-fold from previously,
and made memory usage and data loading dynamic so that it will not use more
than 80 MB of stack unless it needs to.

7/31/03: IMOD_3-0-13
7/31/03: Added ability to send messages to set movie mode and to load a model
without changing black/white levels.

7/31/03: Changed the call used to iconify dialog boxes along with the info
or model view window because the old way didn't work under RedHat 9 or on the 
SGI under Qt 3.1.

7/31/03: Fixed remaining problems created by new treatment of "views", and

mrc/frommrc.c
/*  IMOD VERSION 2.02
 *
 *  frommrc.c -- Convert mrc files to SGI rgb files.
 *
 *  Author: James Kremer email: kremer@colorado.edu
 */

/*****************************************************************************
 *   Copyright (C) 1995-1996 by Boulder Laboratory for 3-Dimensional Fine    *
 *   Structure ("BL3DFS") and the Regents of the University of Colorado.     *
 *                                                                           *
 *   BL3DFS reserves the exclusive rights of preparing derivative works,     *
 *   distributing copies for sale, lease or lending and displaying this      *
 *   software and documentation.                                             *
 *   Users may reproduce the software and documentation as long as the       *
 *   copyright notice and other notices are preserved.                       *
 *   Neither the software nor the documentation may be distributed for       *
 *   profit, either in original form or in derivative works.                 *
 *                                                                           *
 *   THIS SOFTWARE AND/OR DOCUMENTATION IS PROVIDED WITH NO WARRANTY,        *
 *   EXPRESS OR IMPLIED, INCLUDING, WITHOUT LIMITATION, WARRANTY OF          *
 *   MERCHANTABILITY AND WARRANTY OF FITNESS FOR A PARTICULAR PURPOSE.       *
 *                                                                           *
 *   This work is supported by NIH biotechnology grant #RR00592,             *
 *   for the Boulder Laboratory for 3-Dimensional Fine Structure.            *
 *   University of Colorado, MCDB Box 347, Boulder, CO 80309                 *
 *****************************************************************************/
/*  $Author$

    $Date$

    $Revision$

    $Log$
*/

#include <stdio.h>
#include <stdlib.h>

#ifndef __sgi
main(int argc, char **argv)
{
     fprintf(stderr, "%s only runs on sgi machines.\n", argv[0]);
}
#else

#include <gl/image.h>
#include <mrcfiles.h>

putbyte(outf,val)
     FILE *outf;
     unsigned char val;
{
     unsigned char buf[1];
     
     buf[0] = val;
     if (!fwrite(buf,1,1,outf))
	  fprintf(stderr, "putbyte: write error.\n");
     return 0;
}

putshort(outf,val)
     FILE *outf;
     unsigned short val;
{
     unsigned char buf[2];
     
     buf[0] = (val>>8);
     buf[1] = (val>>0);
     fwrite(buf,2,1,outf);
     return 0;
}

static int putlong(outf,val)
     FILE *outf;
     unsigned long val;
{
     unsigned char buf[4];
     
     buf[0] = (val>>24);
     buf[1] = (val>>16);
     buf[2] = (val>>8);
     buf[3] = (val>>0);
     return fwrite(buf,4,1,outf);
}


void mrc_rgb_fillbuf(short *buf, unsigned char *idata, int xsize)
{
     int i;

     for (i = 0; i < xsize; i++)
	  buf[i] = (short)idata[i];
}

main(int argc, char *argv[])
{
     IMAGE *image;
     FILE *fin, *of;
     struct MRCheader hdata;
     short xsize, ysize, zsize;
     int xysize;
     char iname[80];
     unsigned char **idata;
     short *buf;
     int i, x, y, z;
     

     if (argc != 3){
	  fprintf(stderr, 
		  "%s version 1.0\n",  argv[0]);
	  imodCopyright();
	  fprintf(stderr, "%s [mrc file] [rgb name/root]\n\n", argv[0]);
	  fprintf(stderr, "A series of rgb files will be created ");
	  fprintf(stderr, "with the prefix [rgb root name]\n");
	  fprintf(stderr, "and with the suffex nnn.rgb, ");
	  fprintf(stderr, "where nnn is the z number.\n");
	  exit(1);
     }

     if (NULL == (fin = fopen(argv[1], "r"))){
	  fprintf(stderr, "%s: Couldn't open %s\n", argv[0], argv[1]);
	  exit(-1);
     }

     if (mrc_head_read(fin, &hdata)){
	  fprintf(stderr, "%s: Can't Read Input Header from %s.\n",
		  argv[0], argv[1]);
	  exit(-1);
     }

     if (hdata.mode == MRC_MODE_RGB){
	  mrc_rgb_write(fin, &hdata, argv[2]);
	  fclose(fin);
	  exit(0);
     }

     if (hdata.mode){
	  fprintf(stderr, "mrc2rgb: Only byte mode data supported.\n");
	  exit(-1);
     }
     
     xsize = hdata.nx;
     ysize = hdata.ny;
     zsize = hdata.nz;
     xysize = xsize * ysize;

     if (mrc_head_read(fin, &hdata)){
	  fprintf(stderr, "mrc2rgb: Can't Read Input Header from %s.\n",
		  argv[2]);
	  exit(-1);
     }

     idata = mrc_read_byte(fin, &hdata, NULL, NULL);
     
     buf = (short *)malloc(xsize * sizeof(short));

     printf("Writing rgb images. ");
     for (z = 0; z < zsize; z++){
	  
	  printf(".");
	  fflush(stdout);

	  sprintf(iname, "%s.%3.3d.rgb", argv[2], z);
	  if (zsize == 1)
	       sprintf(iname, "%s", argv[2]);

	  image = iopen(iname, "w", RLE(1), 2, xsize, ysize, 1);

	  for(y = 0; y < ysize; y++){
	       mrc_rgb_fillbuf(buf, &(idata[z][y * xsize]), xsize);
	       putrow(image, buf, y, 0);
	  }
	  iclose(image);
     }
     printf("\r\n");
}


mrc_rgb_write(FILE *fin, struct MRCheader *hd, char *rgb)
{
     IMAGE *image;
     unsigned char *rdat, *gdat, *bdat;
     short *buf;
     int xysize;
     int xsize, ysize;
     int z, i, y;
     char iname[128];
     FILE *of;

     xsize = hd->nx;
     ysize = hd->ny;
     xysize = hd->nx * hd->ny;

     rdat = (unsigned char *)malloc(xysize);
     gdat = (unsigned char *)malloc(xysize);
     bdat = (unsigned char *)malloc(xysize);
     buf  = (short *)malloc(xsize * sizeof(short));

     if ( (!rdat) || (!gdat) || (!bdat) || (!buf) ){
	  fprintf(stderr, "mrc_rgb_write: memory alloc error.\n");
	  exit(-1);
     }

     printf("Writing rgb images. ");
     rewind(fin);
     fseek(fin, 1024, SEEK_SET);

     for(z = 0; z < hd->nz; z++){
	  
	  printf(".");
	  fflush(stdout);
	  
	  sprintf(iname, "%s.%3.3d.rgb", rgb, z);
	  if (hd->nz == 1)
	       sprintf(iname, "%s", rgb);
	  image = iopen(iname, "w", RLE(1), 3, xsize, ysize, 3);

	  for(i = 0; i < xysize; i++){
	       fread(&(rdat[i]), 1, 1, fin);
	       fread(&(gdat[i]), 1, 1, fin);
	       fread(&(bdat[i]), 1, 1, fin);
	  }

	  for(y = 0; y < ysize; y++){
	       mrc_rgb_fillbuf(buf, &(rdat[y * xsize]), xsize);
	       putrow(image, buf, y, 0);
	       mrc_rgb_fillbuf(buf, &(gdat[y * xsize]), xsize);
	       putrow(image, buf, y, 1);
	       mrc_rgb_fillbuf(buf, &(bdat[y * xsize]), xsize);
	       putrow(image, buf, y, 2);
	  }

	  iclose(image);
     }
     printf("\r\n");

     free(rdat);
     free(gdat);
     free(bdat);
     free(buf);
     return 0;
}

#endif
